x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -461,4 +608,43 @@ private boolean checkState(IndexerState currentState) {
         }
     }
 
+    private synchronized void reQueueThrottledSearch() {
+        currentMaximumRequestsPerSecond = getMaximumRequestsPerSecond();
+
+        if (scheduledNextSearch != null) {
+            TimeValue executionDelay = calculateThrottlingDelay(
+                currentMaximumRequestsPerSecond,
+                lastDocCount,
+                lastSearchStartTimeNanos,
+                getTimeNanos()
+            );
+
+            logger.debug(
+                "rethrottling job [{}], wait for {} ({} {})",
+                getJobId(),
+                executionDelay,
+                currentMaximumRequestsPerSecond,
+                lastDocCount
+            );
+            scheduledNextSearch.reschedule(executionDelay);
+        }
+    }
+
+    static TimeValue calculateThrottlingDelay(float requestsPerSecond, long docCount, long startTimeNanos, long now) {
+        if (requestsPerSecond <= 0) {
+            return TimeValue.ZERO;
+        }
+        float timeToWaitNanos = (docCount / requestsPerSecond) * TimeUnit.SECONDS.toNanos(1);
```

Comment:
This formula implies that `requestsPerSecond` is really `desiredDocsPerSecond`.  If that's correct then `requestsPerSecond` seems like it will cause confusion in the future because I would assume `requestsPerSecond` referred to the number of searches, each of which could return many documents.

For example, if I saw a configuration parameter `requests_per_second` I might decide to set it to 2 so that I'd get a maximum of 2 search requests per second from this functionality.  But then if one of my searches returns 1000 documents then I get a 500 second wait until the next search.

Reply:
> I saw a couple of nits but I stopped reviewing when I saw the throttling formula because it seems to highlight a fundamental issue: are we intending to throttle based on search requests per second or documents retrieved per second? What changes are required will depend on the answer to that.

This design and the reasons for that are discussed in the tracking issue: #54862. In a nutshell: You are right that `requests_per_second` is misleading. The name has still been chosen, because its a existing concept from `reindex`. It's wrong there, too.

Reply:
Ah yes, I missed that `requests_per_second` is used in this way for reindex already.  In that case consistency with reindex is more important.